azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,122 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr, PriorityQueue
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_telemetry_service_handler(protocol_util):
+    return TelemetryServiceHandler(protocol_util)
+
+
+class TelemetryServiceHandler(ThreadHandlerInterface):
+    """
+    This Handler takes care of sending all telemetry out of the agent to Wireserver. It sends out data as soon as
+    there's any data available in the queue to send.
+    """
+
+    _THREAD_NAME = "TelemetryServiceHandler"
+    _MAX_TIMEOUT = datetime.timedelta(minutes=5).seconds
+
+    def __init__(self, protocol_util):
+        self._protocol = protocol_util.get_protocol()
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = PriorityQueue()
+
+    @staticmethod
+    def get_thread_name():
+        return TelemetryServiceHandler._THREAD_NAME
+
+    def run(self):
+        logger.info("Start Extension Telemetry service.")
+        self.start()
+
+    def is_alive(self):
+        return self._thread is not None and self._thread.is_alive()
+
+    def start(self):
+        self._thread = threading.Thread(target=self._process_telemetry_thread)
+        self._thread.setDaemon(True)
+        self._thread.setName(self.get_thread_name())
+        self._thread.start()
+
+    def stop(self):
+        """
+        Stop server communication and join the thread to main thread.
+        """
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def join(self):
+        self._queue.join()
+        self._thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def enqueue_event_func(self, event):
+        # Add event to queue and set event
+        self._queue.put(event)
+
+        # Set the event if any enqueue happens (even if already set) to trigger sending those events
+        self._should_process_events.set()
+
+    def _process_telemetry_thread(self):
+        logger.info("Successfully started the {0} thread".format(self.get_thread_name()))
+        try:
+            # On demand wait, start processing as soon as there is any data available in the queue
```

Comment:
But we need to wait a little or we could end up sending a network request per event, wouldn't we?

Reply:
There will be a small chance of that, because by the time we get the request ready, another event would mostly be filled in the queue. Even if by the time our first request is ready and we dont have another event, a telemetry_event call to wireserver takes ~2-3secs which is ample amount of time to fill up the queue so that subsequent request calls would have enough data to go around. 

This was another reason to add a generator to yield events from the queue rather than saving all the events in a list and then sending them out one by one.

Reply:
I agree with @narrieta here. I'm not sure if a wait here is the answer, but I definitely think there needs to be some sort of limiter on this code path.

Reply:
+1. Why did we abandon the batch sending approach? Wouldn't we want to keep that + eventually have QuickLog to send out important events on-demand?

Reply:
So now I'm adding back a somewhat hybrid approach, where we batch for either 5secs or 30events, whichever happens first. This is to promote batching and isolation of the telemetry service both at the same time.

Reply:
Done